{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COMP 364 - Computer Tools for Life Sciences\n",
    "## Homework #1\n",
    "Instructors: Christopher J.F. Cameron and Carlos G. Oliver\n",
    "\n",
    "**Due:  Friday, September 29th 2017 at 11:59:59 pm**\n",
    "\n",
    "In this assignment, you will learn how to implement a rudimentry profile or position weight matrix (PWM) from DNA sequences and then scan a separate DNA sequence for possible motif hits. This kind of approach is used frequently to learn what kinds of DNA sequences are likely to be attached to certain proteins.\n",
    "\n",
    "## Computational hints\n",
    "\n",
    "Python programming visualizer: http://www.pythontutor.com/visualize.html#mode=edit\n",
    "\n",
    "All programming concepts are drawn from lectures up to and including 'Putting it together: Nested loops and conditionals'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biology motivation/context\n",
    "\n",
    "From last week's lectures, we now know that there is a transfer of genetic information that occurs within cells from DNA-->RNA-->protein (called the central dogma of biology). In reality, this process is much more complicated as described in the figure below:\n",
    "\n",
    "<img src=\"images/central_dogma.png\" width=\"300\">\n",
    "<image source: https://www.quora.com/What-is-the-central-dogma-of-molecular-biology-Is-it-true (September 16th, 2017)>\n",
    "\n",
    "If you are unfamiliar with the central dogma, we would recommend you review these documents:\n",
    "\n",
    "https://en.wikipedia.org/wiki/Central_dogma_of_molecular_biology\n",
    "\n",
    "https://www.khanacademy.org/test-prep/mcat/biomolecules/amino-acids-and-proteins1/v/central-dogma-of-molecular-biology-2\n",
    "\n",
    "To help control the act of transcription (the transference of genetic information from DNA to RNA), cells employ a set of proteins (called transcription factors) that are known to bind near genes and regulate their expression (e.g., the rate at which they are transcribed into RNA). We are able to observe the activity of transcription factors by using a technology called chromatin immunoprecipitation (or ChIP). This technology uses a targeted antibody to select for DNA sequences that are bound by transcription factors (see figure below).\n",
    "\n",
    "<img src=\"images/chip_tech.png\" width=\"300\">\n",
    "<image adapted from: https://www.researchgate.net/figure/222384404_fig1_Fig-1-Schematic-representation-of-the-chromatin-immunoprecipitation-procedure-The (September 16th, 2017)>\n",
    "\n",
    "The minute (or low level) details about ChIP technology are outside the scope of this course, but you must understand that the result of a ChIP protocol is a collection of DNA sequences that were bound by a specific protein of interest (i.e., a transcription factor).\n",
    "\n",
    "If you would like to learn more about ChIP technology, please see the following links:\n",
    "\n",
    "https://en.wikipedia.org/wiki/ChIP-sequencing\n",
    "\n",
    "http://bitesizebio.com/13541/an-introduction-to-chip-seq/\n",
    "\n",
    "Now the problem we have to address is that we have a collection of DNA sequences that were bound by a protein of interest, but we don't know the specific location where the protein attached to DNA. This is called the motif discovery problem, where:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Given a collection of DNA sequences that were bound by a protein of interest, identify the motif (or binding site) of the protein.\n",
    "\n",
    "Some of you may already be familiar with motifs and have seen examples visualized in a sequence logo:\n",
    "\n",
    "<img src=\"images/motif.png\" width=\"350\">\n",
    "<image source: https://openi.nlm.nih.gov/detailedresult.php?img=PMC3239296_1471-213X-11-59-3&req=4 (September 16th, 2017)>\n",
    "\n",
    "A sequence logo represents of a stack of letters (DNA nucleotides) at each position of the motif (or binding site). The relative sizes of the nucleotides indicate their frequency of representation in the collection of DNA sequences obtained from the ChIP protocol. While the total height of the letters depicts the information content of the position, in bits (a unit that you all should be familiar with now).\n",
    "\n",
    "In the upper-left panel of the figure above, the binding motif of E-box is described as being 11 nucleotides long. We see that positions 6, 7, 8, and 11 have to be nucleotides C, A, G, and G, respectively. While other positions can be multiple nucleotides or have no information about them.\n",
    "\n",
    "A profile or position weight matrix (PWM) is another way to represent the binding site of a protein that was observed to be bound to DNA. In this assignment, we will implement a PWM using only loops, lists/tuples, conditional statements, and minor arithmetic.\n",
    "\n",
    "Let's start by remembering how to declare an empty list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# definition of an empty list\n",
    "empty_list = [] # or list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's do something a little more useful (and biologically more relevant). Define a list object, called 'nucleotides', that contains single character representations of the four possible DNA nucleotides ('A', 'C', 'G', 'T')."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# define list of DNA nucleotides\n",
    "nucleotides = [\"A\",\"C\",\"G\",\"T\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With a list of nucleotides defined, we can now implement some concepts that were introduced in class. \n",
    "\n",
    "First, let's create a loop (either for or while) that iterates over each item in the 'nucleotides' list and prints the item's value & index. \n",
    "\n",
    "a) Using the built-in enumerate function ** (2.5 marks) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 A\n",
      "1 C\n",
      "2 G\n",
      "3 T\n"
     ]
    }
   ],
   "source": [
    "# create a loop that iterates over each item in the list and prints out the item's index & value using enumerate()\n",
    "for index, nucleotide in enumerate(nucleotides):\n",
    "    print(index, nucleotide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Using the built-in index method ** (2.5 marks) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 A\n",
      "1 C\n",
      "2 G\n",
      "3 T\n"
     ]
    }
   ],
   "source": [
    "# create a loop that iterates over each item in the list and prints out the item's index & value using index()\n",
    "for nucleotide in nucleotides:\n",
    "    index = nucleotides.index(nucleotide)\n",
    "    print(index, nucleotide)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's pretend we have just performed a ChIP protocol and observed the following six DNA sequences to be bound by our protein of interest (protein 'X'). Real ChIP-seq DNA sequences can be much longer, but for practicality, we'll limit them to jus six nucleotides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "editable": false
   },
   "outputs": [],
   "source": [
    "# list of observed sequences where transcription factor binds\n",
    "ChIP_seqs = [ \"ACGTGA\", \"ACGGGA\", \"ACGCCC\", \"ACAGAT\", \"GATAGA\", \"ACGTTA\" ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now want to determine the frequency at which each nucleotide is bound at a given position in the motif. Remember, each sequence represents a single observation of the protein's binding motif/site. How many positions are there in the above motif? (complete the code below) ** (2.5 marks) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "motif size: 6\n"
     ]
    }
   ],
   "source": [
    "motif_size = len(ChIP_seqs[0]) #works for any other number within len(ChIP_seqs)-1\n",
    "print(\"motif size:\",motif_size)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each position in the motif, we need to count the frequency of occurence for each possible nucleotides. We're going to start with just the first column and then you will implement a pair of nested loops to complete the matrix. ** (10 marks) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nucleotide counts for position one: [5, 0, 1, 0]\n"
     ]
    }
   ],
   "source": [
    "# declare an empty list for the position weight matrix (PWM) of observed sequences\n",
    "PWM = []\n",
    "\n",
    "# define a list that contains the frequency counts for each nucleotide in the first column\n",
    "nucleotide_counts = [0, 0, 0, 0] # [A, C, G, T] - note: the order is the same as our nucleotides list object above\n",
    "\n",
    "# iterate over each DNA sequence in the list of observed DNA sequences\n",
    "for DNA_seq in ChIP_seqs:\n",
    "    for index, nuc in enumerate(nucleotides):\n",
    "        if nuc == DNA_seq[0]:\n",
    "            nucleotide_counts[index]+= 1\n",
    "\n",
    "# print out the nucleotide counts\n",
    "print(\"nucleotide counts for position one: \" + str(nucleotide_counts))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will expand our code and apply it to all positions of the motif (hint: you will need a loop). ** (10 marks) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[5, 0, 1, 0], [1, 5, 0, 0], [1, 0, 4, 1], [1, 1, 2, 2], [1, 1, 3, 1], [4, 1, 0, 1]]\n"
     ]
    }
   ],
   "source": [
    "PFM = []\n",
    "nucleotide_counts = [0,0,0,0]\n",
    "i = 0\n",
    "while i < motif_size:\n",
    "    for DNA_seq in ChIP_seqs:\n",
    "        for index, nuc in enumerate(nucleotides):\n",
    "            if nuc == DNA_seq[i]:\n",
    "                nucleotide_counts[index]+=1  \n",
    "    PFM.append(nucleotide_counts)\n",
    "    nucleotide_counts = [0,0,0,0] #reset this count after finishing up with each position\n",
    "    i+=1\n",
    "print(PFM)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you have implemented the above code correctly, you should have a two-dimensional list (or a list of lists) containing frequency counts for each nucleotide at a given position in the motif. We would like to represent this in a useful table format such as:\n",
    "\n",
    "Position | A | C | G | T |\n",
    ":---: | :---: | :---: | :---: | :---:\n",
    "1 | $A_{1}$ | $C_{1}$ | $G_{1}$ | $T_{1}$ \n",
    "2 | $A_{2}$ | $C_{2}$ | $G_{2}$ | $T_{2}$ \n",
    "3 | $A_{3}$ | $C_{3}$ | $G_{3}$ | $T_{3}$ \n",
    "4 | $A_{4}$ | $C_{4}$ | $G_{4}$ | $T_{4}$ \n",
    "5 | $A_{5}$ | $C_{5}$ | $G_{5}$ | $T_{5}$ \n",
    "6 | $A_{6}$ | $C_{6}$ | $G_{6}$ | $T_{6}$ \n",
    "\n",
    "Where $A_{1}$ is the frequency count of A at position 1, $A_{2}$ is the frequency count of A at position 2, and so on.\n",
    "\n",
    "This table is called the absolute frequency matrix or position frequency matrix (PFM).\n",
    "\n",
    "To do this, we will write a loop that iterates over the two-dimensional list and prints each position's information. Note: Your table should be informative and easy to read (but it doesn't have to be perfect - i.e., don't waste your time making it look too fancy). ** (10 marks) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pos   A C G T\n",
      "1     5 0 1 0 \n",
      "2     1 5 0 0 \n",
      "3     1 0 4 1 \n",
      "4     1 1 2 2 \n",
      "5     1 1 3 1 \n",
      "6     4 1 0 1 \n"
     ]
    }
   ],
   "source": [
    "print(\"Pos   A C G T\")\n",
    "for index, i in enumerate(PFM):\n",
    "    print(index+1, end=\"     \") #comment out this line and the first print statement for pasting for sequence logo \n",
    "    for j in i:\n",
    "        print(j, end=\" \")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What should the nucleotide frequency counts in each row of your table sum to? **(2.5 marks)**\n",
    "\n",
    "<strong>RESPONSE:</strong> Each row of the table should sum to 6.\n",
    "\n",
    "### Sequence logo creation (5 marks)\n",
    "\n",
    "Go to: http://ccg.vital-it.ch/pwmtools/pwmscan.php#\n",
    "\n",
    "Follow the instructions on the webpage and create a sequence logo of your PFM. All that should be required of you is the following:\n",
    "\n",
    "* set custom weight matrix format\n",
    "* paste your PFM into the text field\n",
    "* click submit\n",
    "\n",
    "Once the page loads you should see a sequence logo of your PFM on the right-side of the screen. Right-click and save the image in the COMP364 HW1's subfolder called 'images'.\n",
    "\n",
    "Create a Markdown box below this text or insert beneath this text the sequence logo using the following command:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`<`img src=\"images/your_file_name.png\" width=\"250\"`>`\n",
    "\n",
    "Replace 'your_file_name.png' with your sequence logo's filename."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3> Sequence Logo: </h3>\n",
    "\n",
    "<img src=\"images/sequence_logo.png\" width=\"250\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Position probability matrices (PPM)\n",
    "\n",
    "Typically, PWMs are calculated as the log-likelihood ratio matrix, where $A_{1}$ represents the log-ratio of the relative frequency to the background frequency for nucleotide 'A' at position 1.\n",
    "\n",
    "First, we'll start by calculating the relative frequencies and creating the position probability matrix (PPM). To do this we will convert the absolute frequency counts to probabilities.\n",
    "\n",
    "How do we convert absolute frequencies to probabilites?\n",
    "\n",
    "Let's define some mathematical notation:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$M$ will be the PPM\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$M_{i,j}$ is the probability of nucleotide $i$ occuring at position $j$ of the PPM\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$N$ is the number of DNA sequences observed to be bound by the protein of interest\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$X_{k,j}$ is the observed nucleotide at position $j$ of observed DNA sequence $k$ \n",
    "\n",
    "By using this notation, we can define $M_{i,j}$ as a function of $X_{k,j}$ such that:\n",
    "\n",
    "$$M_{i,j}=\\frac{1}{N}\\sum_{k=1}^{N}I\\left(X_{k,j}=i\\right)$$\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Where $I()$ returns 1 if $X_{k,j}=i$ and 0 otherwise\n",
    "\n",
    "For anyone unfamiliar with summation notation, sigma ($\\sum$) is the standard for writing long sums. For example, let's sum all of the digits present in the decimal system.\n",
    "\n",
    "$$0+1+2+3+4+5+6+7+8+9=45$$\n",
    "\n",
    "This can be represented as:\n",
    "\n",
    "$$\\sum_{i=0}^{9}i=0+1+2+3+4+5+6+7+8+9=45$$\n",
    "\n",
    "'$\\sum_{i=0}^{9}i$' denotes that we sum all $i$ starting at $i=0$, incrememnt by 1, until we reach $i=9$.\n",
    "\n",
    "If we wanted to represent this as a for loop in Python, it would be:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "editable": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45\n"
     ]
    }
   ],
   "source": [
    "total = 0 # why are we not naming the variable 'sum'?\n",
    "for i in range(10):\n",
    "    total += i\n",
    "print(total)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In practice (and as you will find), implementing the conversion from absolute frequency to a probability is relatively easy. Complete the code below to create a PPM (can be accomplished with as little as two lines of code). ** (10 marks) **\n",
    "\n",
    "Tip - use round() to limit the number of digits after the decimal point when printing the table. See the following link for more information: https://docs.python.org/3.6/library/functions.html#round"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pos     A    C   G   T\n",
      "1     0.83 0.0 0.17 0.0 \n",
      "2     0.17 0.83 0.0 0.0 \n",
      "3     0.17 0.0 0.67 0.17 \n",
      "4     0.17 0.17 0.33 0.33 \n",
      "5     0.17 0.17 0.5 0.17 \n",
      "6     0.67 0.17 0.0 0.17 \n"
     ]
    }
   ],
   "source": [
    "# insert PPM code here\n",
    "PPM = []\n",
    "for position in PFM:\n",
    "    temp = []\n",
    "    for freq in position:\n",
    "        temp.append(freq/len(ChIP_seqs))\n",
    "    PPM.append(temp)\n",
    "    \n",
    "# insert your table printing code below to ensure that your PPM is correct\n",
    "print(\"Pos     A    C   G   T\")\n",
    "for index, i in enumerate(PPM):\n",
    "    print(index+1, end=\"     \") \n",
    "    for j in i:\n",
    "        print(round(j,2), end=\" \")\n",
    "    print()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What should each row of your table sum to now? **(2.5 marks)**\n",
    "\n",
    "<strong> RESPONSE: </strong> The sum of each table should be 1.0 (give or take with rounding).\n",
    "\n",
    "### Likelihoods\n",
    "\n",
    "Now that we have a PPM calculated, we can define the likelihood of observing a specific DNA sequences within the motif. Or in other words, if we give you a DNA sequence of six nucleotides, what is the probability that the protein of interest will bind it?\n",
    "\n",
    "For example, we observed the DNA sequence 'GTAGTA'. The likelihood, $L$, of of this sequence being bound by our protein of interest is:\n",
    "\n",
    "$$ L(seq\\ |\\ M) = M_{G,1} \\times M_{T,2} \\times M_{A,3} \\times M_{G,4} \\times M_{T,5} \\times M_{A,6} $$\n",
    "\n",
    "What is the likelihood of observing the following sequences? **(1 mark each for correct answers) ** \n",
    "\n",
    "a) CGTAGT = <strong>0.0</strong>\n",
    "\n",
    "b) ACGTAA = <strong>0.017146776406035666</strong>\n",
    "\n",
    "c) AAAAAA = <strong>0.00042866941015089156</strong>\n",
    "\n",
    "(hint: fill in the code below to detemine the likelihood of observing the above sequences) ** (7 marks for correct implementation) **"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "likelihood of 'AAAAAA': 0.00042866941015089156\n"
     ]
    }
   ],
   "source": [
    "# define sequence to calculate likelihood of\n",
    "sequence = \"AAAAAA\" # for example\n",
    "\n",
    "# define initial likelihood\n",
    "likelihood = 1.0\n",
    "\n",
    "# insert likelihood calculation code here\n",
    "likelihoods = []\n",
    "for index, letter in enumerate(sequence):\n",
    "    if letter == \"A\":\n",
    "        likelihoods.append(PPM[index][0])\n",
    "    elif letter == \"C\":\n",
    "        likelihoods.append(PPM[index][1])\n",
    "    elif letter == \"G\":\n",
    "        likelihoods.append(PPM[index][2])\n",
    "    elif letter == \"T\":\n",
    "        likelihoods.append(PPM[index][3])\n",
    "    else:\n",
    "        print(\"Invalid nucleotide.\")\n",
    "        break\n",
    "for number in likelihoods:\n",
    "    likelihood*= number\n",
    "#likelihood = round(likelihood,1)   #if you would like to round \n",
    "print(\"likelihood of '\"+sequence+\"':\",likelihood)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a useful program to calculate likelihoods that you have created, but it's not very practical. Add user functionality to the code: **(10 marks)**\n",
    "\n",
    "* continually prompt the user for a string containing a DNA sequence\n",
    "* ensure that the provided string is valid to have the likelihood calculated (print a useful error message if it does not)\n",
    "* then calculate the likelihood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please paste your DNA sequence: AAAAAA\n",
      "Likelihood of 'AAAAAA': 0.00042866941015089156\n",
      "Quit? (Y/N) Y\n",
      "\n",
      "Your session has ended.\n"
     ]
    }
   ],
   "source": [
    "# define sequence to calculate likelihood of\n",
    "answer = True\n",
    "while answer == True:\n",
    "    sequence = input(\"Please paste your DNA sequence: \") # for example\n",
    "\n",
    "    # define initial likelihood\n",
    "    likelihood = 1.0\n",
    "\n",
    "    # insert likelihood calculation code here\n",
    "    likelihoods = []\n",
    "    sequence = sequence.strip() #removes white space from input\n",
    "    #checks if the sequence is the correct length to use with this PPM\n",
    "    if len(sequence) == len(PPM):\n",
    "        #check if all letters in the sequence are in fact valid nucleotides\n",
    "        for index, letter in enumerate(sequence):\n",
    "            if letter == \"A\":\n",
    "                likelihoods.append(PPM[index][0])\n",
    "            elif letter == \"C\":\n",
    "                likelihoods.append(PPM[index][1])\n",
    "            elif letter == \"G\":\n",
    "                likelihoods.append(PPM[index][2])\n",
    "            elif letter == \"T\":\n",
    "                likelihoods.append(PPM[index][3])\n",
    "            else:\n",
    "                print(\"Invalid nucleotide present. Please try again.\")\n",
    "                answer = False\n",
    "                break\n",
    "        if answer == True:\n",
    "            for number in likelihoods:\n",
    "                likelihood*= number\n",
    "            print(\"Likelihood of '\"+sequence+\"':\", likelihood)\n",
    "            #ask for response again if invalid\n",
    "            again = True\n",
    "            while again == True:\n",
    "                response = input(\"Quit? (Y/N) \")\n",
    "                print() #spacing for legibility\n",
    "                if response.lower() == \"n\":\n",
    "                    answer = True\n",
    "                    again = False\n",
    "                elif response.lower() == \"y\":            \n",
    "                    print(\"Your session has ended.\")\n",
    "                    answer = False\n",
    "                    again = False\n",
    "                else:\n",
    "                    print(\"Invalid response. Please try again.\")\n",
    "                    again = True\n",
    "        else: #if nucleotide is invalid, I want it to ask again, so must set to true again so while loop check works\n",
    "            answer = True\n",
    "    else:\n",
    "        print(\"Your DNA sequence is too long or short for this PPM. Please try again.\")\n",
    "        answer = True\n",
    "        continue;\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Position weight matrices (PWM)\n",
    "\n",
    "Finally, we will now convert our PPM to a PWM by calculating the log2-ratio between the observed probabilities (found in the PPM) and the background probability. For simplicity, we will assume each nucleotide occurs with an equal probability or:\n",
    "\n",
    "$$p_{i}=\\frac{1}{|i|}$$\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Where $|i|$ is the size of our nucleotide alphabet.\n",
    "\n",
    "What should the background frequncy be for any given nucleotide? **(2.5 marks)**\n",
    "\n",
    "<strong>RESPONSE: </strong> The background frequency for any given nucleotide should be 0.25.\n",
    "\n",
    "Now we will calculate the PWM by calculating the log2-ratio between the PPM's probabilities and the background probability, where:\n",
    "\n",
    "$$ PWM_{i,j} = log_{2}(\\frac{M_{i,j}}{p_{i}})$$\n",
    "\n",
    "\n",
    "Again this is relatively easy and can be done in a few lines of code. **( 10 marks) **\n",
    "\n",
    "To help with the log2 transformation, we will import the Python math module (more on modules in the future) and use the math.log2() function. This function accepts a numerical object as input and returns the base-2 logarithm applied to the input. See the following link for more information: https://docs.python.org/3.4/library/math.html\n",
    "\n",
    "You may encounter problems when taking the log2() of some ratios (i.e., log2(0) is negative infinity and zero divided by any number is undefined). This can be avoided by either using conditional statements or adding +1 to each ratio (feel free to choose either)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pos     A    C   G   T\n",
      "1     2.12 0.0 0.74 0.0 \n",
      "2     0.74 2.12 0.0 0.0 \n",
      "3     0.74 0.0 1.87 0.74 \n",
      "4     0.74 0.74 1.22 1.22 \n",
      "5     0.74 0.74 1.58 0.74 \n",
      "6     1.87 0.74 0.0 0.74 \n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "PWM = []\n",
    "#PWM created earlier but will do again\n",
    "# insert PWM code here\n",
    "for item in PPM:\n",
    "    temp = []\n",
    "    for i in item:\n",
    "        ratio = (i/0.25)+1\n",
    "        calc = math.log2(ratio)\n",
    "        temp.append(calc)\n",
    "    PWM.append(temp)\n",
    "\n",
    "\n",
    "# insert your table printing code below to ensure that your PWM is correct\n",
    "\n",
    "print(\"Pos     A    C   G   T\")\n",
    "for index, i in enumerate(PWM):\n",
    "    print(index+1, end=\"     \") \n",
    "    for j in i:\n",
    "        print(round(j,2), end=\" \")\n",
    "    print()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Motif scanning\n",
    "\n",
    "Another use for PWMs is to search for occurrences of the motif in a given DNA sequence. To do this we use what is called a sliding window approach.\n",
    "\n",
    "For example, we observe a DNA sequence of size 10 using a window of size 3.\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DNA sequence = ACGTACTATT\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Windows of size 3 = [ACG], [CGT], [GTA], [TAC], [CTA], [TAT], [ATT]\n",
    "\n",
    "In this context, the first instance of the sliding window will observe nucleotides 0-2, the second instance will then observe nucleotides 1-3, and so on. The window continues to slide along the DNA sequence until it reaches the end.\n",
    "\n",
    "For each window, we will calculate a score (similar to the likelihood) that we define as the following:\n",
    "\n",
    "$$score_{s} = \\sum_{j=1}^{N}PWM_{s,j}$$\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Where $s$ is the observed nucleotide in the DNA sequence located at position $j$ of the PWM.\n",
    " \n",
    "Given the defined list object below containing a random DNA sequence, identify all posible locations (above a given score threshold) where the protein may bind and sore hits based on the log2-ratio. Provide the total number of hits as well as the DNA index for the start and score of each window. ** (15 marks) **\n",
    "\n",
    "For example:\n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;total number of hits: \n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DNA index: ??\t(score) \n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DNA index: ??\t(score) \n",
    "\n",
    "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DNA index: ??\t(score) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "editable": false
   },
   "outputs": [],
   "source": [
    "# DNA sequence to be scanned\n",
    "DNA_sequence = \"GTGTCGCGACAGCCCAGCATGATATCCTGAGGCGCTACGCCGACGGATGTCCCACGGAATTGCCACGGTAGCCGAGCGCTACGCGGGCGACACGAACTTATGTATGGAGCGGGCCGTCGAAAGGTCGTACCCTTGCAGTTAGCACGTAGCCCGGCCCCACTAGCACAGCAGTGCCTCGGGCGGCATCCTCATTATTAAGTTTTCTCTACAGCCAAACGACCAGGTGCGCTTCCGCGGAGCGCGGTGGAGACTCGCCCACCCGGCAGCTCTGTGACGGGGACTAAAGGGGCGATGATAATCGCGAGTGCCGCGTTATGGTGGTGTCGGGACAGAGCGGCCTTGCGGCCAGTCGTATCCCTTCTCGAGCTCCGTCCGGTTAAGCGTGACACCCCCAGCGTACCCGCAAACCGCGATGGCTGTGCTTGGGGTCGATCGCACGTAGGACGGTCCCCAACGTGAGACGCCGGGGCACCAGTTCTCACGCCCAAAGCATAAACGACGGGCAG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of hits:  237\n",
      "Motif: TCGCGA DNA Index: 298 (8.19)\n",
      "Motif: GCGACA DNA Index: 86 (8.08)\n",
      "Motif: ACAGCC DNA Index: 207 (7.66)\n",
      "Motif: AGCCCA DNA Index: 10 (5.46)\n",
      "Motif: CCCAGC DNA Index: 390 (5.17)\n",
      "Motif: CCAGCA DNA Index: 13 (6.69)\n",
      "Motif: GCATGA DNA Index: 16 (8.27)\n",
      "Motif: ATGATA DNA Index: 291 (7.34)\n",
      "Motif: ATATCC DNA Index: 21 (5.55)\n",
      "Motif: TCCTGA DNA Index: 24 (6.8)\n",
      "Motif: GAGGCG DNA Index: 28 (5.31)\n",
      "Motif: AGGCGC DNA Index: 29 (7.05)\n",
      "Motif: GCGCTA DNA Index: 75 (8.08)\n",
      "Motif: GCTACG DNA Index: 77 (5.06)\n",
      "Motif: ACGCCG DNA Index: 460 (7.58)\n",
      "Motif: GCCGAC DNA Index: 38 (5.55)\n",
      "Motif: CCGACG DNA Index: 39 (5.46)\n",
      "Motif: GACGGA DNA Index: 41 (6.16)\n",
      "Motif: ACGGAT DNA Index: 42 (8.8)\n",
      "Motif: GGATGT DNA Index: 44 (5.02)\n",
      "Motif: ATGTCC DNA Index: 46 (6.69)\n",
      "Motif: CCACGG DNA Index: 62 (5.17)\n",
      "Motif: CACGGA DNA Index: 52 (5.42)\n",
      "Motif: ACGGAA DNA Index: 53 (9.94)\n",
      "Motif: AATTGC DNA Index: 57 (7.13)\n",
      "Motif: ATTGCC DNA Index: 58 (5.55)\n",
      "Motif: TTGCCA DNA Index: 59 (5.22)\n",
      "Motif: CCACGG DNA Index: 62 (5.17)\n",
      "Motif: ACGGTA DNA Index: 64 (9.94)\n",
      "Motif: AGCCGA DNA Index: 69 (6.31)\n",
      "Motif: CCGAGC DNA Index: 71 (7.05)\n",
      "Motif: GAGCGC DNA Index: 236 (6.41)\n",
      "Motif: GCGCTA DNA Index: 75 (8.08)\n",
      "Motif: GCTACG DNA Index: 77 (5.06)\n",
      "Motif: ACGCGG DNA Index: 80 (8.43)\n",
      "Motif: GCGGGC DNA Index: 108 (8.27)\n",
      "Motif: GGGCGA DNA Index: 286 (6.81)\n",
      "Motif: GCGACA DNA Index: 86 (8.08)\n",
      "Motif: ACACGA DNA Index: 89 (9.16)\n",
      "Motif: ACGAAC DNA Index: 91 (8.32)\n",
      "Motif: AACTTA DNA Index: 94 (6.69)\n",
      "Motif: ACTTAT DNA Index: 95 (7.66)\n",
      "Motif: TATGTA DNA Index: 98 (5.31)\n",
      "Motif: ATGTAT DNA Index: 99 (6.69)\n",
      "Motif: TATGGA DNA Index: 102 (6.16)\n",
      "Motif: ATGGAG DNA Index: 103 (5.95)\n",
      "Motif: GAGCGG DNA Index: 331 (5.67)\n",
      "Motif: GCGGGC DNA Index: 108 (8.27)\n",
      "Motif: GCCGTC DNA Index: 112 (5.55)\n",
      "Motif: CCGTCG DNA Index: 113 (5.95)\n",
      "Motif: TCGAAA DNA Index: 116 (7.34)\n",
      "Motif: AAAGGT DNA Index: 119 (7.13)\n",
      "Motif: AAGGTC DNA Index: 120 (7.42)\n",
      "Motif: AGGTCG DNA Index: 121 (5.95)\n",
      "Motif: TCGTAC DNA Index: 124 (6.69)\n",
      "Motif: ACCCTT DNA Index: 128 (6.44)\n",
      "Motif: CCTTGC DNA Index: 337 (6.4)\n",
      "Motif: GCAGTT DNA Index: 134 (6.29)\n",
      "Motif: CAGTTA DNA Index: 135 (6.45)\n",
      "Motif: GCACGT DNA Index: 434 (6.65)\n",
      "Motif: ACGTAG DNA Index: 436 (8.06)\n",
      "Motif: GCCCGG DNA Index: 148 (5.17)\n",
      "Motif: CCCGGC DNA Index: 258 (5.66)\n",
      "Motif: CCGGCC DNA Index: 150 (6.69)\n",
      "Motif: GCCCCA DNA Index: 153 (6.2)\n",
      "Motif: CCACTA DNA Index: 156 (6.2)\n",
      "Motif: ACTAGC DNA Index: 158 (8.03)\n",
      "Motif: AGCACA DNA Index: 161 (5.46)\n",
      "Motif: GCACAG DNA Index: 162 (5.06)\n",
      "Motif: ACAGCA DNA Index: 164 (8.8)\n",
      "Motif: AGCAGT DNA Index: 166 (5.17)\n",
      "Motif: GCAGTG DNA Index: 167 (5.55)\n",
      "Motif: CAGTGC DNA Index: 168 (6.16)\n",
      "Motif: AGTGCC DNA Index: 303 (5.55)\n",
      "Motif: CCTCGG DNA Index: 173 (5.17)\n",
      "Motif: TCGGGC DNA Index: 175 (7.53)\n",
      "Motif: GCGGCA DNA Index: 179 (8.56)\n",
      "Motif: GCATCC DNA Index: 182 (6.29)\n",
      "Motif: TCCTCA DNA Index: 185 (5.95)\n",
      "Motif: CCTCAT DNA Index: 186 (5.06)\n",
      "Motif: TCATTA DNA Index: 188 (6.69)\n",
      "Motif: ATTATT DNA Index: 190 (5.06)\n",
      "Motif: TATTAA DNA Index: 192 (5.31)\n",
      "Motif: AAGTTT DNA Index: 196 (7.42)\n",
      "Motif: AGTTTT DNA Index: 197 (5.55)\n",
      "Motif: TCTCTA DNA Index: 202 (6.2)\n",
      "Motif: TCTACA DNA Index: 204 (6.2)\n",
      "Motif: ACAGCC DNA Index: 207 (7.66)\n",
      "Motif: CAGCCA DNA Index: 208 (5.96)\n",
      "Motif: AGCCAA DNA Index: 209 (5.46)\n",
      "Motif: GCCAAA DNA Index: 210 (6.2)\n",
      "Motif: CCAAAC DNA Index: 211 (5.06)\n",
      "Motif: AAACGA DNA Index: 493 (7.79)\n",
      "Motif: AACGAC DNA Index: 494 (5.55)\n",
      "Motif: ACGACC DNA Index: 215 (8.32)\n",
      "Motif: ACCAGG DNA Index: 218 (6.55)\n",
      "Motif: CCAGGT DNA Index: 219 (6.4)\n",
      "Motif: AGGTGC DNA Index: 221 (7.53)\n",
      "Motif: GTGCGC DNA Index: 223 (5.67)\n",
      "Motif: GCGCTT DNA Index: 225 (6.94)\n",
      "Motif: GCTTCC DNA Index: 227 (6.29)\n",
      "Motif: CCGCGG DNA Index: 231 (6.31)\n",
      "Motif: GCGGAG DNA Index: 233 (6.69)\n",
      "Motif: GAGCGC DNA Index: 236 (6.41)\n",
      "Motif: GCGCGG DNA Index: 238 (7.05)\n",
      "Motif: GCGGTG DNA Index: 240 (6.69)\n",
      "Motif: GGTGGA DNA Index: 242 (6.16)\n",
      "Motif: TGGAGA DNA Index: 244 (6.07)\n",
      "Motif: GAGACT DNA Index: 246 (5.56)\n",
      "Motif: AGACTC DNA Index: 247 (5.06)\n",
      "Motif: ACTCGC DNA Index: 249 (8.03)\n",
      "Motif: TCGCCC DNA Index: 251 (6.2)\n",
      "Motif: GCCCAC DNA Index: 253 (5.06)\n",
      "Motif: CCACCC DNA Index: 255 (5.06)\n",
      "Motif: ACCCGG DNA Index: 257 (6.55)\n",
      "Motif: CCCGGC DNA Index: 258 (5.66)\n",
      "Motif: CCGGCA DNA Index: 259 (7.82)\n",
      "Motif: GCAGCT DNA Index: 262 (6.29)\n",
      "Motif: GCTCTG DNA Index: 265 (5.06)\n",
      "Motif: CTGTGA DNA Index: 268 (6.56)\n",
      "Motif: ACGGGG DNA Index: 273 (8.91)\n",
      "Motif: CGGGGA DNA Index: 274 (6.56)\n",
      "Motif: GGGGAC DNA Index: 275 (5.31)\n",
      "Motif: GACTAA DNA Index: 278 (5.31)\n",
      "Motif: ACTAAA DNA Index: 279 (8.32)\n",
      "Motif: AAAGGG DNA Index: 282 (6.4)\n",
      "Motif: AAGGGG DNA Index: 283 (7.53)\n",
      "Motif: AGGGGC DNA Index: 284 (7.53)\n",
      "Motif: GGGCGA DNA Index: 286 (6.81)\n",
      "Motif: GCGATG DNA Index: 409 (6.2)\n",
      "Motif: CGATGA DNA Index: 289 (5.42)\n",
      "Motif: ATGATA DNA Index: 291 (7.34)\n",
      "Motif: ATAATC DNA Index: 294 (5.06)\n",
      "Motif: AATCGC DNA Index: 296 (6.65)\n",
      "Motif: TCGCGA DNA Index: 298 (8.19)\n",
      "Motif: GCGAGT DNA Index: 300 (7.79)\n",
      "Motif: GAGTGC DNA Index: 302 (6.89)\n",
      "Motif: AGTGCC DNA Index: 303 (5.55)\n",
      "Motif: CCGCGT DNA Index: 307 (7.05)\n",
      "Motif: GCGTTA DNA Index: 309 (8.56)\n",
      "Motif: TATGGT DNA Index: 313 (5.02)\n",
      "Motif: ATGGTG DNA Index: 314 (5.95)\n",
      "Motif: GGTGGT DNA Index: 316 (5.02)\n",
      "Motif: TGGTGT DNA Index: 318 (5.42)\n",
      "Motif: TCGGGA DNA Index: 323 (8.67)\n",
      "Motif: GGGACA DNA Index: 325 (5.96)\n",
      "Motif: GACAGA DNA Index: 327 (5.67)\n",
      "Motif: ACAGAG DNA Index: 328 (6.93)\n",
      "Motif: CAGAGC DNA Index: 329 (5.67)\n",
      "Motif: GAGCGG DNA Index: 331 (5.67)\n",
      "Motif: AGCGGC DNA Index: 332 (5.66)\n",
      "Motif: GCGGCC DNA Index: 341 (7.42)\n",
      "Motif: CCTTGC DNA Index: 337 (6.4)\n",
      "Motif: GCGGCC DNA Index: 341 (7.42)\n",
      "Motif: CGGCCA DNA Index: 342 (5.22)\n",
      "Motif: GCCAGT DNA Index: 344 (5.91)\n",
      "Motif: CCAGTC DNA Index: 345 (5.55)\n",
      "Motif: AGTCGT DNA Index: 347 (5.91)\n",
      "Motif: TCGTAT DNA Index: 349 (6.69)\n",
      "Motif: CCTTCT DNA Index: 356 (5.55)\n",
      "Motif: TCTCGA DNA Index: 359 (7.05)\n",
      "Motif: TCGAGC DNA Index: 361 (7.05)\n",
      "Motif: GAGCTC DNA Index: 363 (5.56)\n",
      "Motif: GCTCCG DNA Index: 365 (5.06)\n",
      "Motif: CCGTCC DNA Index: 368 (6.69)\n",
      "Motif: TCCGGT DNA Index: 371 (5.66)\n",
      "Motif: CCGGTT DNA Index: 372 (6.69)\n",
      "Motif: CGGTTA DNA Index: 373 (5.71)\n",
      "Motif: GGTTAA DNA Index: 374 (5.31)\n",
      "Motif: AAGCGT DNA Index: 378 (7.79)\n",
      "Motif: GCGTGA DNA Index: 380 (9.41)\n",
      "Motif: GTGACA DNA Index: 382 (5.96)\n",
      "Motif: ACACCC DNA Index: 385 (7.18)\n",
      "Motif: ACCCCC DNA Index: 387 (6.44)\n",
      "Motif: CCCCCA DNA Index: 388 (5.46)\n",
      "Motif: CCCAGC DNA Index: 390 (5.17)\n",
      "Motif: CAGCGT DNA Index: 392 (5.67)\n",
      "Motif: AGCGTA DNA Index: 393 (5.95)\n",
      "Motif: GCGTAC DNA Index: 394 (7.42)\n",
      "Motif: ACCCGC DNA Index: 398 (7.29)\n",
      "Motif: CCCGCA DNA Index: 399 (5.95)\n",
      "Motif: CCGCAA DNA Index: 400 (7.34)\n",
      "Motif: GCAAAC DNA Index: 402 (5.8)\n",
      "Motif: AAACCG DNA Index: 404 (5.06)\n",
      "Motif: AACCGC DNA Index: 405 (5.91)\n",
      "Motif: ACCGCG DNA Index: 406 (6.19)\n",
      "Motif: CCGCGA DNA Index: 407 (8.19)\n",
      "Motif: GCGATG DNA Index: 409 (6.2)\n",
      "Motif: GATGGC DNA Index: 411 (5.76)\n",
      "Motif: ATGGCT DNA Index: 412 (6.69)\n",
      "Motif: GCTGTG DNA Index: 415 (5.55)\n",
      "Motif: CTGTGC DNA Index: 416 (5.42)\n",
      "Motif: GCTTGG DNA Index: 420 (6.4)\n",
      "Motif: TGGGGT DNA Index: 423 (5.42)\n",
      "Motif: GGGGTC DNA Index: 424 (5.31)\n",
      "Motif: GGTCGA DNA Index: 426 (5.67)\n",
      "Motif: TCGATC DNA Index: 428 (6.2)\n",
      "Motif: GATCGC DNA Index: 430 (5.27)\n",
      "Motif: ATCGCA DNA Index: 431 (5.95)\n",
      "Motif: TCGCAC DNA Index: 432 (6.2)\n",
      "Motif: GCACGT DNA Index: 434 (6.65)\n",
      "Motif: ACGTAG DNA Index: 436 (8.06)\n",
      "Motif: GTAGGA DNA Index: 438 (6.16)\n",
      "Motif: TAGGAC DNA Index: 439 (5.31)\n",
      "Motif: AGGACG DNA Index: 440 (5.46)\n",
      "Motif: GACGGT DNA Index: 442 (5.02)\n",
      "Motif: ACGGTC DNA Index: 443 (8.8)\n",
      "Motif: TCCCCA DNA Index: 447 (5.46)\n",
      "Motif: CCCCAA DNA Index: 448 (5.46)\n",
      "Motif: ACGTGA DNA Index: 453 (10.79)\n",
      "Motif: GTGAGA DNA Index: 455 (6.81)\n",
      "Motif: AGACGC DNA Index: 458 (5.91)\n",
      "Motif: ACGCCG DNA Index: 460 (7.58)\n",
      "Motif: GCCGGG DNA Index: 462 (5.66)\n",
      "Motif: CCGGGG DNA Index: 463 (6.8)\n",
      "Motif: CGGGGC DNA Index: 464 (5.42)\n",
      "Motif: GGGGCA DNA Index: 465 (6.45)\n",
      "Motif: GCACCA DNA Index: 468 (6.94)\n",
      "Motif: ACCAGT DNA Index: 470 (7.29)\n",
      "Motif: CCAGTT DNA Index: 471 (5.55)\n",
      "Motif: CAGTTC DNA Index: 472 (5.31)\n",
      "Motif: AGTTCT DNA Index: 473 (5.55)\n",
      "Motif: TCTCAC DNA Index: 476 (5.06)\n",
      "Motif: TCACGC DNA Index: 478 (5.91)\n",
      "Motif: ACGCCC DNA Index: 480 (8.32)\n",
      "Motif: GCCCAA DNA Index: 482 (6.2)\n",
      "Motif: CCCAAA DNA Index: 483 (5.46)\n",
      "Motif: AAAGCA DNA Index: 486 (7.42)\n",
      "Motif: AAGCAT DNA Index: 487 (6.94)\n",
      "Motif: AGCATA DNA Index: 488 (5.46)\n",
      "Motif: GCATAA DNA Index: 489 (7.42)\n",
      "Motif: ATAAAC DNA Index: 491 (5.06)\n",
      "Motif: AAACGA DNA Index: 493 (7.79)\n",
      "Motif: AACGAC DNA Index: 494 (5.55)\n",
      "Motif: ACGACG DNA Index: 495 (7.58)\n",
      "Motif: ACGGGC DNA Index: 498 (9.65)\n",
      "Motif: CGGGCA DNA Index: 499 (5.71)\n"
     ]
    }
   ],
   "source": [
    "# define empty list to store motif hits\n",
    "hits = []\n",
    "scores = []\n",
    "hit_threshold = 5.0 # example value, try different values and see the effect\n",
    "PWMs = []\n",
    "DNA_indices = {}\n",
    "desc = {} \n",
    "# insert sliding window approach here\n",
    "window_size = 6 #since we have 6 positions in our PWM\n",
    "for i in range(0,len(DNA_sequence)-window_size):\n",
    "    score = 0\n",
    "    temp = DNA_sequence[i:i+window_size]\n",
    "    for position, nuc in enumerate(temp):\n",
    "        if nuc == \"A\":\n",
    "            PWMs.append(PWM[position][0])\n",
    "        elif nuc == \"C\":\n",
    "            PWMs.append(PWM[position][1])\n",
    "        elif nuc == \"G\":\n",
    "            PWMs.append(PWM[position][2])\n",
    "        elif nuc == \"T\":\n",
    "            PWMs.append(PWM[position][3])\n",
    "        else:\n",
    "            print(\"Invalid nucleotide.\")\n",
    "    #sum up PWMs to score\n",
    "    for item in PWMs:\n",
    "        score += item\n",
    "    PWMs = [] #reset PWMs since I'm using append()\n",
    "    if score > hit_threshold:\n",
    "        hits.append(temp)\n",
    "        scores.append(score)\n",
    "        desc[temp] = score #score been stored here too so so scores list can be overridden later\n",
    "        DNA_indices[temp] = i\n",
    "\n",
    "print (\"Total number of hits: \",len(hits))\n",
    "for sequence in hits:\n",
    "    print(\"Motif:\", sequence, \"DNA Index:\", DNA_indices[sequence], \"(\" + str(round(desc[sequence],2)) + \")\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**BONUS (+5 marks)**: sort motif hits in descending order based on their score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of hits:  237\n",
      "Motif: ACGTGA DNA Index: 453 (10.79)\n",
      "Motif: ACGGTA DNA Index: 64 (9.94)\n",
      "Motif: ACGGAA DNA Index: 53 (9.94)\n",
      "Motif: ACGGGC DNA Index: 498 (9.65)\n",
      "Motif: GCGTGA DNA Index: 380 (9.41)\n",
      "Motif: ACACGA DNA Index: 89 (9.16)\n",
      "Motif: ACGGGG DNA Index: 273 (8.91)\n",
      "Motif: ACGGTC DNA Index: 443 (8.8)\n",
      "Motif: ACGGAT DNA Index: 42 (8.8)\n",
      "Motif: ACAGCA DNA Index: 164 (8.8)\n",
      "Motif: TCGGGA DNA Index: 323 (8.67)\n",
      "Motif: GCGTTA DNA Index: 309 (8.56)\n",
      "Motif: GCGGCA DNA Index: 179 (8.56)\n",
      "Motif: ACGCGG DNA Index: 80 (8.43)\n",
      "Motif: ACGACC DNA Index: 215 (8.32)\n",
      "Motif: ACGAAC DNA Index: 91 (8.32)\n",
      "Motif: ACGCCC DNA Index: 480 (8.32)\n",
      "Motif: ACTAAA DNA Index: 279 (8.32)\n",
      "Motif: GCGGGC DNA Index: 108 (8.27)\n",
      "Motif: GCGGGC DNA Index: 108 (8.27)\n",
      "Motif: GCATGA DNA Index: 16 (8.27)\n",
      "Motif: TCGCGA DNA Index: 298 (8.19)\n",
      "Motif: CCGCGA DNA Index: 407 (8.19)\n",
      "Motif: TCGCGA DNA Index: 298 (8.19)\n",
      "Motif: GCGCTA DNA Index: 75 (8.08)\n",
      "Motif: GCGACA DNA Index: 86 (8.08)\n",
      "Motif: GCGACA DNA Index: 86 (8.08)\n",
      "Motif: GCGCTA DNA Index: 75 (8.08)\n",
      "Motif: ACGTAG DNA Index: 436 (8.06)\n",
      "Motif: ACGTAG DNA Index: 436 (8.06)\n",
      "Motif: ACTAGC DNA Index: 158 (8.03)\n",
      "Motif: ACTCGC DNA Index: 249 (8.03)\n",
      "Motif: CCGGCA DNA Index: 259 (7.82)\n",
      "Motif: AAACGA DNA Index: 493 (7.79)\n",
      "Motif: AAACGA DNA Index: 493 (7.79)\n",
      "Motif: GCGAGT DNA Index: 300 (7.79)\n",
      "Motif: AAGCGT DNA Index: 378 (7.79)\n",
      "Motif: ACAGCC DNA Index: 207 (7.66)\n",
      "Motif: ACAGCC DNA Index: 207 (7.66)\n",
      "Motif: ACTTAT DNA Index: 95 (7.66)\n",
      "Motif: ACGCCG DNA Index: 460 (7.58)\n",
      "Motif: ACGACG DNA Index: 495 (7.58)\n",
      "Motif: ACGCCG DNA Index: 460 (7.58)\n",
      "Motif: AAGGGG DNA Index: 283 (7.53)\n",
      "Motif: AGGTGC DNA Index: 221 (7.53)\n",
      "Motif: TCGGGC DNA Index: 175 (7.53)\n",
      "Motif: AGGGGC DNA Index: 284 (7.53)\n",
      "Motif: GCGTAC DNA Index: 394 (7.42)\n",
      "Motif: GCGGCC DNA Index: 341 (7.42)\n",
      "Motif: AAAGCA DNA Index: 486 (7.42)\n",
      "Motif: GCATAA DNA Index: 489 (7.42)\n",
      "Motif: GCGGCC DNA Index: 341 (7.42)\n",
      "Motif: AAGGTC DNA Index: 120 (7.42)\n",
      "Motif: AAGTTT DNA Index: 196 (7.42)\n",
      "Motif: CCGCAA DNA Index: 400 (7.34)\n",
      "Motif: TCGAAA DNA Index: 116 (7.34)\n",
      "Motif: ATGATA DNA Index: 291 (7.34)\n",
      "Motif: ATGATA DNA Index: 291 (7.34)\n",
      "Motif: ACCCGC DNA Index: 398 (7.29)\n",
      "Motif: ACCAGT DNA Index: 470 (7.29)\n",
      "Motif: ACACCC DNA Index: 385 (7.18)\n",
      "Motif: AATTGC DNA Index: 57 (7.13)\n",
      "Motif: AAAGGT DNA Index: 119 (7.13)\n",
      "Motif: GCGCGG DNA Index: 238 (7.05)\n",
      "Motif: TCTCGA DNA Index: 359 (7.05)\n",
      "Motif: AGGCGC DNA Index: 29 (7.05)\n",
      "Motif: TCGAGC DNA Index: 361 (7.05)\n",
      "Motif: CCGCGT DNA Index: 307 (7.05)\n",
      "Motif: CCGAGC DNA Index: 71 (7.05)\n",
      "Motif: GCACCA DNA Index: 468 (6.94)\n",
      "Motif: GCGCTT DNA Index: 225 (6.94)\n",
      "Motif: AAGCAT DNA Index: 487 (6.94)\n",
      "Motif: ACAGAG DNA Index: 328 (6.93)\n",
      "Motif: GAGTGC DNA Index: 302 (6.89)\n",
      "Motif: GGGCGA DNA Index: 286 (6.81)\n",
      "Motif: GTGAGA DNA Index: 455 (6.81)\n",
      "Motif: GGGCGA DNA Index: 286 (6.81)\n",
      "Motif: TCCTGA DNA Index: 24 (6.8)\n",
      "Motif: CCGGGG DNA Index: 463 (6.8)\n",
      "Motif: TCATTA DNA Index: 188 (6.69)\n",
      "Motif: GCGGAG DNA Index: 233 (6.69)\n",
      "Motif: CCAGCA DNA Index: 13 (6.69)\n",
      "Motif: GCGGTG DNA Index: 240 (6.69)\n",
      "Motif: AACTTA DNA Index: 94 (6.69)\n",
      "Motif: ATGGCT DNA Index: 412 (6.69)\n",
      "Motif: TCGTAC DNA Index: 124 (6.69)\n",
      "Motif: ATGTCC DNA Index: 46 (6.69)\n",
      "Motif: TCGTAT DNA Index: 349 (6.69)\n",
      "Motif: CCGGCC DNA Index: 150 (6.69)\n",
      "Motif: ATGTAT DNA Index: 99 (6.69)\n",
      "Motif: CCGTCC DNA Index: 368 (6.69)\n",
      "Motif: CCGGTT DNA Index: 372 (6.69)\n",
      "Motif: GCACGT DNA Index: 434 (6.65)\n",
      "Motif: AATCGC DNA Index: 296 (6.65)\n",
      "Motif: GCACGT DNA Index: 434 (6.65)\n",
      "Motif: CTGTGA DNA Index: 268 (6.56)\n",
      "Motif: CGGGGA DNA Index: 274 (6.56)\n",
      "Motif: ACCAGG DNA Index: 218 (6.55)\n",
      "Motif: ACCCGG DNA Index: 257 (6.55)\n",
      "Motif: GGGGCA DNA Index: 465 (6.45)\n",
      "Motif: CAGTTA DNA Index: 135 (6.45)\n",
      "Motif: ACCCTT DNA Index: 128 (6.44)\n",
      "Motif: ACCCCC DNA Index: 387 (6.44)\n",
      "Motif: GAGCGC DNA Index: 236 (6.41)\n",
      "Motif: GAGCGC DNA Index: 236 (6.41)\n",
      "Motif: AAAGGG DNA Index: 282 (6.4)\n",
      "Motif: CCAGGT DNA Index: 219 (6.4)\n",
      "Motif: GCTTGG DNA Index: 420 (6.4)\n",
      "Motif: CCTTGC DNA Index: 337 (6.4)\n",
      "Motif: CCTTGC DNA Index: 337 (6.4)\n",
      "Motif: AGCCGA DNA Index: 69 (6.31)\n",
      "Motif: CCGCGG DNA Index: 231 (6.31)\n",
      "Motif: GCTTCC DNA Index: 227 (6.29)\n",
      "Motif: GCATCC DNA Index: 182 (6.29)\n",
      "Motif: GCAGTT DNA Index: 134 (6.29)\n",
      "Motif: GCAGCT DNA Index: 262 (6.29)\n",
      "Motif: TCTACA DNA Index: 204 (6.2)\n",
      "Motif: CCACTA DNA Index: 156 (6.2)\n",
      "Motif: GCCAAA DNA Index: 210 (6.2)\n",
      "Motif: GCCCCA DNA Index: 153 (6.2)\n",
      "Motif: TCTCTA DNA Index: 202 (6.2)\n",
      "Motif: GCCCAA DNA Index: 482 (6.2)\n",
      "Motif: GCGATG DNA Index: 409 (6.2)\n",
      "Motif: GCGATG DNA Index: 409 (6.2)\n",
      "Motif: TCGCCC DNA Index: 251 (6.2)\n",
      "Motif: TCGATC DNA Index: 428 (6.2)\n",
      "Motif: TCGCAC DNA Index: 432 (6.2)\n",
      "Motif: ACCGCG DNA Index: 406 (6.19)\n",
      "Motif: GGTGGA DNA Index: 242 (6.16)\n",
      "Motif: TATGGA DNA Index: 102 (6.16)\n",
      "Motif: GACGGA DNA Index: 41 (6.16)\n",
      "Motif: GTAGGA DNA Index: 438 (6.16)\n",
      "Motif: CAGTGC DNA Index: 168 (6.16)\n",
      "Motif: TGGAGA DNA Index: 244 (6.07)\n",
      "Motif: CAGCCA DNA Index: 208 (5.96)\n",
      "Motif: GTGACA DNA Index: 382 (5.96)\n",
      "Motif: GGGACA DNA Index: 325 (5.96)\n",
      "Motif: AGCGTA DNA Index: 393 (5.95)\n",
      "Motif: CCCGCA DNA Index: 399 (5.95)\n",
      "Motif: TCCTCA DNA Index: 185 (5.95)\n",
      "Motif: ATCGCA DNA Index: 431 (5.95)\n",
      "Motif: AGGTCG DNA Index: 121 (5.95)\n",
      "Motif: CCGTCG DNA Index: 113 (5.95)\n",
      "Motif: ATGGAG DNA Index: 103 (5.95)\n",
      "Motif: ATGGTG DNA Index: 314 (5.95)\n",
      "Motif: AACCGC DNA Index: 405 (5.91)\n",
      "Motif: TCACGC DNA Index: 478 (5.91)\n",
      "Motif: AGTCGT DNA Index: 347 (5.91)\n",
      "Motif: GCCAGT DNA Index: 344 (5.91)\n",
      "Motif: AGACGC DNA Index: 458 (5.91)\n",
      "Motif: GCAAAC DNA Index: 402 (5.8)\n",
      "Motif: GATGGC DNA Index: 411 (5.76)\n",
      "Motif: CGGTTA DNA Index: 373 (5.71)\n",
      "Motif: CGGGCA DNA Index: 499 (5.71)\n",
      "Motif: CAGCGT DNA Index: 392 (5.67)\n",
      "Motif: GACAGA DNA Index: 327 (5.67)\n",
      "Motif: GGTCGA DNA Index: 426 (5.67)\n",
      "Motif: GTGCGC DNA Index: 223 (5.67)\n",
      "Motif: CAGAGC DNA Index: 329 (5.67)\n",
      "Motif: GAGCGG DNA Index: 331 (5.67)\n",
      "Motif: GAGCGG DNA Index: 331 (5.67)\n",
      "Motif: GCCGGG DNA Index: 462 (5.66)\n",
      "Motif: CCCGGC DNA Index: 258 (5.66)\n",
      "Motif: AGCGGC DNA Index: 332 (5.66)\n",
      "Motif: TCCGGT DNA Index: 371 (5.66)\n",
      "Motif: CCCGGC DNA Index: 258 (5.66)\n",
      "Motif: GAGACT DNA Index: 246 (5.56)\n",
      "Motif: GAGCTC DNA Index: 363 (5.56)\n",
      "Motif: CCAGTC DNA Index: 345 (5.55)\n",
      "Motif: AGTTTT DNA Index: 197 (5.55)\n",
      "Motif: CCAGTT DNA Index: 471 (5.55)\n",
      "Motif: AGTTCT DNA Index: 473 (5.55)\n",
      "Motif: GCCGTC DNA Index: 112 (5.55)\n",
      "Motif: CCTTCT DNA Index: 356 (5.55)\n",
      "Motif: GCCGAC DNA Index: 38 (5.55)\n",
      "Motif: AGTGCC DNA Index: 303 (5.55)\n",
      "Motif: ATATCC DNA Index: 21 (5.55)\n",
      "Motif: GCTGTG DNA Index: 415 (5.55)\n",
      "Motif: GCAGTG DNA Index: 167 (5.55)\n",
      "Motif: AACGAC DNA Index: 494 (5.55)\n",
      "Motif: ATTGCC DNA Index: 58 (5.55)\n",
      "Motif: AGTGCC DNA Index: 303 (5.55)\n",
      "Motif: AACGAC DNA Index: 494 (5.55)\n",
      "Motif: AGCCAA DNA Index: 209 (5.46)\n",
      "Motif: AGCATA DNA Index: 488 (5.46)\n",
      "Motif: TCCCCA DNA Index: 447 (5.46)\n",
      "Motif: CCCCAA DNA Index: 448 (5.46)\n",
      "Motif: CCCCCA DNA Index: 388 (5.46)\n",
      "Motif: AGCACA DNA Index: 161 (5.46)\n",
      "Motif: CCCAAA DNA Index: 483 (5.46)\n",
      "Motif: AGCCCA DNA Index: 10 (5.46)\n",
      "Motif: AGGACG DNA Index: 440 (5.46)\n",
      "Motif: CCGACG DNA Index: 39 (5.46)\n",
      "Motif: CTGTGC DNA Index: 416 (5.42)\n",
      "Motif: CGATGA DNA Index: 289 (5.42)\n",
      "Motif: TGGGGT DNA Index: 423 (5.42)\n",
      "Motif: CACGGA DNA Index: 52 (5.42)\n",
      "Motif: TGGTGT DNA Index: 318 (5.42)\n",
      "Motif: CGGGGC DNA Index: 464 (5.42)\n",
      "Motif: GGTTAA DNA Index: 374 (5.31)\n",
      "Motif: GACTAA DNA Index: 278 (5.31)\n",
      "Motif: GAGGCG DNA Index: 28 (5.31)\n",
      "Motif: TATTAA DNA Index: 192 (5.31)\n",
      "Motif: TATGTA DNA Index: 98 (5.31)\n",
      "Motif: TAGGAC DNA Index: 439 (5.31)\n",
      "Motif: CAGTTC DNA Index: 472 (5.31)\n",
      "Motif: GGGGTC DNA Index: 424 (5.31)\n",
      "Motif: GGGGAC DNA Index: 275 (5.31)\n",
      "Motif: GATCGC DNA Index: 430 (5.27)\n",
      "Motif: TTGCCA DNA Index: 59 (5.22)\n",
      "Motif: CGGCCA DNA Index: 342 (5.22)\n",
      "Motif: CCCAGC DNA Index: 390 (5.17)\n",
      "Motif: AGCAGT DNA Index: 166 (5.17)\n",
      "Motif: GCCCGG DNA Index: 148 (5.17)\n",
      "Motif: CCACGG DNA Index: 62 (5.17)\n",
      "Motif: CCACGG DNA Index: 62 (5.17)\n",
      "Motif: CCCAGC DNA Index: 390 (5.17)\n",
      "Motif: CCTCGG DNA Index: 173 (5.17)\n",
      "Motif: TCTCAC DNA Index: 476 (5.06)\n",
      "Motif: CCACCC DNA Index: 255 (5.06)\n",
      "Motif: AAACCG DNA Index: 404 (5.06)\n",
      "Motif: GCACAG DNA Index: 162 (5.06)\n",
      "Motif: GCTACG DNA Index: 77 (5.06)\n",
      "Motif: CCTCAT DNA Index: 186 (5.06)\n",
      "Motif: GCTACG DNA Index: 77 (5.06)\n",
      "Motif: GCTCCG DNA Index: 365 (5.06)\n",
      "Motif: GCTCTG DNA Index: 265 (5.06)\n",
      "Motif: ATAAAC DNA Index: 491 (5.06)\n",
      "Motif: AGACTC DNA Index: 247 (5.06)\n",
      "Motif: CCAAAC DNA Index: 211 (5.06)\n",
      "Motif: ATTATT DNA Index: 190 (5.06)\n",
      "Motif: ATAATC DNA Index: 294 (5.06)\n",
      "Motif: GCCCAC DNA Index: 253 (5.06)\n",
      "Motif: GACGGT DNA Index: 442 (5.02)\n",
      "Motif: GGTGGT DNA Index: 316 (5.02)\n",
      "Motif: TATGGT DNA Index: 313 (5.02)\n",
      "Motif: GGATGT DNA Index: 44 (5.02)\n"
     ]
    }
   ],
   "source": [
    "#sort motifs by descending score\n",
    "\n",
    "#associate scores with motifs, where score is the keys\n",
    "#sort by comparing index 0 to 1,2,3,....n, and then second loop do 1 to 2,3,4 ...n then 2 to 3,4,5...n\n",
    "\n",
    "i=0\n",
    "while i < len(scores)-1:\n",
    "    j = i+1\n",
    "    while j < len(scores):\n",
    "        if scores[i] < scores[j]:\n",
    "            temp_score = scores[i] #anything I do to scores I must do to hits to keep the same order!\n",
    "            temp_seq = hits[i]\n",
    "            scores[i] = scores[j]\n",
    "            hits[i] = hits[j]\n",
    "            scores[j] = temp_score\n",
    "            hits[j] = temp_seq\n",
    "        j+=1\n",
    "    i+=1 \n",
    "\n",
    "#print out hits to stdout\n",
    "\n",
    "print (\"Total number of hits: \",len(hits))\n",
    "for sequence in hits:\n",
    "   print(\"Motif:\", sequence, \"DNA Index:\", DNA_indices[sequence], \"(\" + str(round(desc[sequence],2)) + \")\")"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
